torch_xla/csrc/ops/dynamic_ir.cpp

```diff
@@ -0,0 +1,118 @@
+#include "torch_xla/csrc/ops/dynamic_ir.h"
+
+#include "absl/strings/str_join.h"
+#include "tensorflow/compiler/xla/xla_client/debug_macros.h"
+#include "torch_xla/csrc/lowering_context.h"
+#include "torch_xla/csrc/ops/infer_output_shape.h"
+
+static const torch::lazy::DimensionNode* DimCast(torch::lazy::Output output) {
+  return dynamic_cast<const torch::lazy::DimensionNode*>(output.node);
+}
+
+namespace torch_xla {
+
+SizeNode::SizeNode(torch::lazy::Value input, size_t dim)
+    : XlaNode(torch::lazy::OpKind{c10::Symbol::fromQualString("aten::size")},
+              {input}, xla::ShapeUtil::MakeShape(xla::S32, {}), 1,
+              torch::lazy::MHash(dim)),
+      dim_(dim){};
+
+XlaOpVector SizeNode::Lower(LoweringContext* loctx) const {
+  auto input = loctx->GetOutputOp(operand(0));
+  return ReturnOp(xla::GetDimensionSize(input, this->dim_), loctx);
+}
+
+int64_t SizeNode::getStaticValue() const {
+  return dynamic_cast<const XlaNode*>(operand(0).node)->shape(0).size(dim_);
+}
+
+bool SizeNode::isDynamic() const {
+  auto symbolic_vec =
+      dynamic_cast<const XlaNode*>(operand(0).node)->shape(0).is_symbolic();
+  if (!symbolic_vec.has_value()) {
+    return true;
+  }
+  return symbolic_vec->at(dim_);
+}
+
+std::string SizeNode::ToString() const { return "SizeNode"; }
+
+SizeAdd::SizeAdd(torch::lazy::Value a, torch::lazy::Value b)
+    : XlaNode(torch::lazy::OpKind{c10::Symbol::fromQualString("aten::add")},
+              {a, b}, xla::ShapeUtil::MakeShape(xla::S32, {}), 1){};
```

Comment:
shouldn't the shape of `SizeAdd` to be input dependent?

Reply:
on a second thought, does `SizeAdd` always applied to a single dimension? If that's the case it will always be a single s32?

Reply:
My understanding is yes.